spec/controllers/holds_controller_spec.rb

```diff
@@ -169,5 +170,218 @@
+      context 'when placing hold to a holdable item that has no volumes' do
+        let(:success_response) {
+          { "holdRecord": { "key": 'a_hold_key' } }
+        }
+
+        before do
+          place_hold_params['barcodes'] = 'a_holdable_barcode'
+          allow(place_hold_response).to receive(:status).and_return 200
+          allow(place_hold_response).to receive(:body).and_return success_response.to_json
+        end
+
+        it 'sets the barcode and the hold key in session' do
+          post :create, params: place_hold_params
+
+          expect(session[:place_hold_results][:success].first).to include(
+            barcode: 'a_holdable_barcode',
+            hold_key: 'a_hold_key'
+          )
+        end
+
+        it 'sets catkey in session' do
+          post :create, params: place_hold_params
+
+          expect(session[:place_hold_catkey]).to eq '1'
+        end
+
+        it 'redirects to result page' do
+          post :create, params: place_hold_params
+
+          expect(response).to redirect_to result_path
+        end
+      end
+
+      context 'when placing hold to a holdable item with volumes' do
+        let(:mock_client) { instance_double(SymphonyClient) }
+        let(:place_hold_response_1) { instance_double(HTTP::Response, status: 200, body: success_response_1) }
+        let(:place_hold_response_2) { instance_double(HTTP::Response, status: 200, body: success_response_2) }
+        let(:place_hold_response_3) { instance_double(HTTP::Response, status: 400, body: error_response) }
+        let(:success_response_1) { { "holdRecord": { "key": 'hold_key_1' } }.to_json }
+        let(:success_response_2) { { "holdRecord": { "key": 'hold_key_2' } }.to_json }
+        let(:error_response) { { "messageList": [{ 'message': 'User already has a hold on this material' }] }.to_json }
+
+        let(:place_hold_results) { {
+          success: [{ barcode: 'holdable_barcode_1', hold_key: 'hold_key_1' },
+                    { barcode: 'holdable_barcode_2', hold_key: 'hold_key_2' }],
+          error: [{ barcode: 'not_holdable_barcode', error_message: 'User already has a hold on this material' }]
+        } }
+
+        before do
+          allow(mock_client).to receive(:place_hold)
+            .and_return(place_hold_response_1, place_hold_response_2, place_hold_response_3)
+          place_hold_params['barcodes'] = ['holdable_barcode_1', 'holdable_barcode_2', 'not_holdable_barcode']
+        end
+
+        it 'sets the barcodes and the hold keys in session' do
+          post :create, params: place_hold_params
+
+          expect(session[:place_hold_results]).to eq place_hold_results
+        end
+
+        it 'sets catkey in session' do
+          post :create, params: place_hold_params
+
+          expect(session[:place_hold_catkey]).to eq '1'
+        end
+
+        it 'redirects to result page' do
+          post :create, params: place_hold_params
+
+          expect(response).to redirect_to result_path
+        end
```

Comment:
just not sure this has to be tested in separate scenarios, right? these two are the same code regardless of permutations of context. right?

Reply:
yeah, this bugged me a bit either actually. I tried to not repeat those by nesting an upper level context but then rubocop nagged about the max # of nested context. Maybe you are right, there is no need to test them again in the other scenarios. so create action will redirect to result_path if there are some results(success or failed), and will redirect to holds_path if results are empty. Anyway I will just leave these checks in one context and remove from others

Reply:
I think you could dump expectations on catkey and redirect in the main context... like before the contexts. Your call though, lmk what you want to do.